# Assignment 1

**Group 22**

A. Siganos ( 1283871 )     
J. Gómez Robles ( 1286552 )

In [1]:
import numpy as np
np.random.seed(13) #TODO Check if this is used for sgd
import keras.backend as K
from keras.models import Sequential
from keras.layers import Dense, Embedding, Reshape, Lambda
from keras.utils import np_utils
from keras.utils.data_utils import get_file
from keras.preprocessing.text import Tokenizer
from keras.utils.vis_utils import model_to_dot
from keras.preprocessing import sequence
from gensim.models import KeyedVectors
from sklearn.manifold import TSNE
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.neighbors import NearestNeighbors as nn
from matplotlib import pylab
from __future__ import division

import time
# Only for the word2vec parsing of last exercise (it was part of the template, not our implementation)
#import gensim


Using TensorFlow backend.


In [2]:
# DO NOT Modify the lines in this cell
path = 'alice.txt'
corpus = open(path).readlines()[0:700]

corpus = [sentence for sentence in corpus if sentence.count(" ") >= 2]

tokenizer = Tokenizer(filters='!"#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n'+"'")
tokenizer.fit_on_texts(corpus)
corpus = tokenizer.texts_to_sequences(corpus)
nb_samples = sum(len(s) for s in corpus)
V = len(tokenizer.word_index) + 1

# Is this something they need to change?
dim = 100
window_size = 2 #use this window size for Skipgram, CBOW, and the model with the additional hidden layer
window_size_corpus = 4 #use this window size for the co-occurrence matrix

## Question 1

### Co-occurrence Matrix
Use the provided code to load the "Alice in Wonderland" text document. 
1. Implement the word-word co-occurrence matrix for “Alice in Wonderland”
2. Normalize the words such that every value lies within a range of 0 and 1
3. Compute the cosine distance between the given words:
    - Alice 
    - Dinah
    - Rabbit
4. List the 5 closest words to 'Alice'. Discuss the results.
5. Discuss what the main drawbacks are of a term-term co-occurence matrix solutions?


In [3]:
################################################################################
############################## 0. Util functions ###############################
################################################################################

# Run this block first
def getKeyFromValue(dictionary, target):
    return list(dictionary.keys())[list(dictionary.values()).index(target)]


In [4]:
################################################################################
################# 1. Implement word-word co-ocurrence matrix ###################
################################################################################

# Create my co-ocurrence matrix, initially 0 (VxV size)
# How to access this matrix:
#   Each column is an (index - 1) (column 0 is word in index 1; column 1 is word in index 2; ...)
wcoMatrix = np.zeros([V, V]) 

# Using my window_size_corpus to define my context scope
scope = window_size_corpus

# TODO: Do not make that many iterations
# Greedy approach first (to be able to compare the optimization)

# For each line in the corpus. Note that they preserve the order, even when they are indexes now.
for s in corpus:
    # For each word
    for current_index in range(0, len(s)):
        current_value = s[current_index] # Get the 'word'
        # From left to right
        for neighbor_index in range(current_index - scope, current_index + scope + 1):
            # Never out of boundaries and never the same index
            if ( neighbor_index >= 0 and neighbor_index < len(s) ) and ( neighbor_index != current_index ):
                # Get my neighbor 'word'
                neighbor = s[neighbor_index] 
                # Never myself and myself (keeping diagonal to 0)
                if current_value != neighbor:
                    # Update the ocurrence
                    #wcoMatrix[current_value - 1, neighbor - 1] += 1
                    wcoMatrix[current_value, neighbor] += 1
                    
################################################################################
################################# 2. Normalize #################################
################################################################################
wcoMatrix = wcoMatrix / wcoMatrix.max()
wcoMatrix

print("Sumary of the word-word co-ocurrence matrix")
print(" - Shape", wcoMatrix.shape)
print(" - Min value", wcoMatrix.min())
print(" - Max value", wcoMatrix.max())


Sumary of the word-word co-ocurrence matrix
 - Shape (1183, 1183)
 - Min value 0.0
 - Max value 1.0


In [56]:
################################################################################
########### 3. Compute cosine similarity to Alice, Dinah and Rabbit ############
################################################################################

words_to_compare = ["Alice", "Dinah", "Rabbit"]

# Get my index
full_index = tokenizer.word_index

# Iterate on all possible combinations
for w1 in words_to_compare:
    w1 = w1.lower() 
    x = full_index[w1]
    for w2 in words_to_compare:
        w2 = w2.lower()
        if w1 != w2:
            y = full_index[w2]
            X = wcoMatrix[x, :].reshape((1, V))
            Y = wcoMatrix[y, :].reshape((1, V))
            print("cosine_similarity(%s, %s)=%s" % (w1, w2, cosine_similarity(X, Y)))
            

cosine_similarity(alice, dinah)=[[0.39360011]]
cosine_similarity(alice, rabbit)=[[0.47890931]]
cosine_similarity(dinah, alice)=[[0.39360011]]
cosine_similarity(dinah, rabbit)=[[0.29862324]]
cosine_similarity(rabbit, alice)=[[0.47890931]]
cosine_similarity(rabbit, dinah)=[[0.29862324]]


In [6]:
################################################################################
################## 4a. List the 5 closest words to 'Alice'. ####################
################################################################################

################################
### We use Nearest Neighbors ###
################################

# It was not importing
from sklearn.neighbors import NearestNeighbors

# Define our variables
targetWord = "Alice".lower()
X = wcoMatrix
N_NEIGHBORS = 5
full_index = tokenizer.word_index

# Perfrom the NearestNeibhbors algorithm
# TODO: Tune parameters
nbrs = NearestNeighbors(n_neighbors = N_NEIGHBORS).fit(X)

# Get the index corresponding to the target word
aliceMatrixIndex = full_index[targetWord]

# Get results from nearest neighbors
distances, indices = nbrs.kneighbors()
aliceIndices = indices[aliceMatrixIndex, :]
aliceDistances = distances[aliceMatrixIndex, :]

# Output result
print("Nearest neighbors (co-ocurrence matrix based) to %s (index %s):" % (targetWord, aliceMatrixIndex))
for i in range(0, N_NEIGHBORS):
    print( "%s [index %s] with distance: %s" % (getKeyFromValue(full_index, aliceIndices[i]), aliceIndices[i], aliceDistances[i]) )

# TODO: Think about what we are computing. 
# With first approach, a neighbor would be a word with a similar vector in the co-ocurrence matrix, 
# not a word within the same context (window_size). Although words in the similar window_size should have a
# similar vector, isn't?

Nearest neighbors (co-ocurrence matrix based) to alice (index 11):
her [index 14] with distance: 0.5616433150108668
that [index 13] with distance: 0.5692975698101006
herself [index 41] with distance: 0.5700573426956416
for [index 18] with distance: 0.5811858638252018
on [index 21] with distance: 0.5837124477057559


**4b. Discuss the results**

We observe that the most similar words are possesives and prepositions. Although these results intuitively make some sense (Alice is the main character of the book and consequently, describing her behaviour is expected), no idiomatic relations can be achieved by means of the simple word-word co-ocurrence matrix.

Furthermore, the minimal distance (cosine similarity normalized) found for alice is of 0.5616, which suggests a poor result - one would be interested on a similarity close to 0.


**5. Discuss what the main drawbacks are of a term-term co-occurence matrix solutions? **

TODO. Add some ideas!

Jorge's ideas: sparse matrix VxV size (computation time increases).



In [7]:
#Save your all the vector representations of your word embeddings in this way
#Change when necessary the sizes of the vocabulary/embedding dimension

f = open('vectors_co_occurrence.txt',"w")
f.write(" ".join([str(V-1),str(V-1)]))
f.write("\n")

#vectors = your word co-occurrence matrix
vectors = wcoMatrix
for word, i in tokenizer.word_index.items():    
    f.write(word)
    f.write(" ")
    f.write(" ".join(map(str, list(vectors[i,:]))))
    f.write("\n")
f.close()


In [8]:
#reopen your file as follows
# TODO: Check! It is not reloading the file D:
# co_occurrence = KeyedVectors.load_word2vec_format('./vectors_co_occurrence.txt', binary=False)


## Question 2

### Word embeddings
Build embeddings with a keras implementation where the embedding vector is of length 50, 150 and 300. Use the Alice in Wonderland text book for training.
1. Using the CBOW model
2. Using Skipgram model
3. Add extra hidden dense layer to CBow and Skipgram implementations. Choose an activation function for that layer and justify your answer.
4. Analyze the four different word embeddings
    - Implement your own function to perform the analogy task with. Do not use existing libraries for this task such as Gensim. Your function should be able to answer whether an anaology as in the example given in the pdf-file is true.
    - Compare the performance on the analogy task between the word embeddings that you have trained in 2.1, 2.2 and 2.3.  
    - Visualize your results and interpret your results
5. Use the word co-occurence matrix from Question 1. Compare the performance on the analogy task with the performance of your trained word embeddings.  
6. Discuss:
    - What are the main advantages of CBOW and Skipgram?
    - What is the advantage of negative sampling?
    - What are the main drawbacks of CBOW and Skipgram?
7. Load pre-trained embeddings on large corpuses (see the pdf file). You only have to consider the word embeddings with an embedding size of 300
    - Compare performance on the analogy task with your own trained embeddings from "Alice in Wonderland". You can limit yourself to the vocabulary of Alice in Wonderland. Visualize the pre-trained word embeddings and compare these with the results of your own trained word embeddings. 


# CBOW

In [51]:
#prepare data for cbow
def generate_data_cbow(corpus, window_size, V):
    all_in = [] 
    all_out = []
    maxlen = window_size*2
    for words in corpus:
        L = len(words)
        for index, word in enumerate(words):
            contexts = []
            labels   = []            
            s = index - window_size
            e = index + window_size + 1
            
            contexts.append([words[i] for i in range(s, e) if 0 <= i < L and i != index])
            labels.append(word)
            if contexts != []:
                all_in.append(sequence.pad_sequences(contexts, maxlen = maxlen))
                all_out.append(np_utils.to_categorical(labels, V))
    return (all_in,all_out) # return the values

            
            
# def generate_data_skipgram(corpus, window_size, V):
#     all_in = [] 
#     all_out = []
#     for sentence in corpus: # For, sentence in corpus
#         L = len(sentence) # Limit for the iteration (per sentence)

#         # For each pair in the index
#         for index, word in enumerate(sentence):
#             # Create X and Y
#             in_words = [] # list of lists
#             labels = [] # list of integers

#             # Iterate over all the sentence, for each word in the line. I.e, for each word:
#             for i in range(index - window_size, index + window_size + 1):
#                 # If between the limits and not myself
#                 if i != index and (0 <= i < L):
#                     # This means: Associating the words in my neighbor (context) 
#                     in_words.append([word]) # Current word (as a list) is associated to...
#                     labels.append(sentence[i]) # ... this word (as a simple integer).
                    
#             # Once I have saved the context for the current word...
#             # ... if there is something in the arrays
#             if in_words != []:
#                 # Save the information to the all_in/out arrays
#                 all_in.append(np.array(in_words,dtype=np.int32)) # X in integer format
#                 # Set Y to categorical values. If my output is ([6, 26]). Then I create
#                 # a vector from 0 to V and mark 6 and 26 as 1
#                 all_out.append(np_utils.to_categorical(labels, V))
#     return (all_in,all_out) # return the values



            

In [52]:
#create CBOW model

def getCbowModel(dim = 50, extra_layer = False, activation_extra_layer = 'softmax'):

    # The Sequential model is a linear stack of layers.
    cbow = Sequential()
    cbow.add(Embedding(input_dim = V, output_dim = dim, input_length = window_size*2))

    # TODO!!! 
    if extra_layer :
        cbow.add(Dense(input_dim = dim, units=V, kernel_initializer='uniform', activation = activation_extra_layer))
    
    cbow.add( Lambda(lambda x: K.mean(x, axis = 1), output_shape = (dim,)) )
    cbow.add( Dense(V, activation = 'softmax') )

    # define loss function for Cbow
    cbow.compile(loss='categorical_crossentropy', optimizer = 'adadelta')

    return cbow


In [54]:
#train model

################################################################################
############################ 1. Using CBOW model ###############################
################################################################################

dims = [50, 150, 300]
iterations = 10
listResX = []
listResY = []

for dimension in dims:
    print("Running CBOW (dim = %s)..."% dimension)
    cbow = getCbowModel(dimension, False)
    X, Y = generate_data_cbow(corpus, window_size, V)
    for ite in range(10):
        loss = 0.
        for x, y in zip(X, Y):
            loss += cbow.train_on_batch(x, y)
        print(ite, loss)

    end_time = time.time()
    
    print("--- %s seconds ---" % (end_time - start_time))
    listResX.append(X)
    listResY.append(Y)


Running CBOW (dim = 50)...
0 41689.845880031586
1 39133.51404154301
2 38928.91506630182
3 38833.24115335941
4 38761.70057722926
5 38703.20775666833
6 38651.24161220342
7 38607.990240603685
8 38575.85596059263
9 38554.49107839912
--- 619.8965148925781 seconds ---
Running CBOW (dim = 150)...
0 41622.75203227997
1 38757.28512895107
2 38349.863362789154


KeyboardInterrupt: 

# Skipgram

In [13]:
#prepare data for Skipgram

# This function creates a list where _each word_ in the text file (with non-empty content)
# is encoded into two arrays:
#   - x: an array of the size of all his windows_size neighbors with the value of the word (i.e, repeated values)
#   - y: an array of the same size as x pointing to each neighbor. Finally encoded in a |V|-list as one-hot enc
def generate_data_skipgram(corpus, window_size, V):
    all_in = [] 
    all_out = []
    for sentence in corpus: # For, sentence in corpus
        L = len(sentence) # Limit for the iteration (per sentence)

        # For each pair in the index
        for index, word in enumerate(sentence):
            # Create X and Y
            in_words = [] # list of lists
            labels = [] # list of integers

            # Iterate over all the sentence, for each word in the line. I.e, for each word:
            for i in range(index - window_size, index + window_size + 1):
                # If between the limits and not myself
                if i != index and (0 <= i < L):
                    # This means: Associating the words in my neighbor (context) 
                    in_words.append([word]) # Current word (as a list) is associated to...
                    labels.append(sentence[i]) # ... this word (as a simple integer).
                    
            # Once I have saved the context for the current word...
            # ... if there is something in the arrays
            if in_words != []:
                # Save the information to the all_in/out arrays
                all_in.append(np.array(in_words,dtype=np.int32)) # X in integer format
                # Set Y to categorical values. If my output is ([6, 26]). Then I create
                # a vector from 0 to V and mark 6 and 26 as 1
                all_out.append(np_utils.to_categorical(labels, V))
    return (all_in,all_out) # return the values

# Prepare data for skipgram
#X, Y = generate_data_skipgram(corpus, window_size, V)


In [14]:
# #save the preprocessed data of Skipgram
# f = open('data_skipgram.txt' ,'w')

# for input, outcome in zip(X,Y):
#     input = np.concatenate(input)
#     f.write(" ".join(map(str, list(input))))
#     f.write(",")
#     outcome = np.concatenate(outcome)
#     f.write(" ".join(map(str,list(outcome))))
#     f.write("\n")
# f.close()


In [15]:
# #load the preprocessed Skipgram data
# def generate_data_skipgram_from_file():
#     f = open('data_skipgram.txt' ,'r')
#     for row in f:
#         inputs,outputs = row.split(",")
#         inputs = np.fromstring(inputs, dtype=int, sep=' ')
#         inputs = np.asarray(np.split(inputs, len(inputs)))
#         outputs = np.fromstring(outputs, dtype=float, sep=' ')
#         outputs = np.asarray(np.split(outputs, len(inputs)))
#         yield (inputs,outputs)
        

In [16]:
# create Skipgram model

def getSkipgramModel(dim = 50, extra_layer = False, activation_extra_layer = 'softmax'):

    # The Sequential model is a linear stack of layers.
    skipgram = Sequential()
    
    # First layer (Projection layer. Embedding for wt)
    # 
    # Parameters: 
    #   - input_dim: V (From the book: 1xV (input layer))
    #   - embeddings_initializer = Glorot uniform. Is it for the uniform distribution of the noises?
    #   - input_length: 1. It is needed for the dense layer (https://keras.io/layers/embeddings/#embedding)
    # Input: One-hot ecoding of word wt in the 1xV space
    # Output: vj in the dimension space 1xd
    skipgram.add( Embedding(input_dim = V, output_dim = dim, embeddings_initializer = 'glorot_uniform', input_length = 1) )

    # Intermediate layer. Can be seen as a helper only.
    skipgram.add( Reshape((dim, )) )

    # Second Layer (Output layer. Probabilities of context words)
    # From documentation, Dense implements the operation: 
    #      output = activation(dot(input, kernel) + bias) 
    #   where:
    #      activation is the element-wise activation function passed as the activation argument
    #      kernel is a weights matrix created by the layer
    #      and bias is a bias vector created by the layer (only applicable if use_bias is True). _WE DO NOT SET THIS_.
    #
    # Hence, we are computing ck*vj
    #
    # Parameters:
    #   - input_dim: dim. From embedding layer, dx1
    #   - units: V. The output dimension, From the book 1xV
    #   - kernel_initializer: uniform (?)
    #   - activation: softmax (as described in the book)
    if extra_layer :
        skipgram.add(Dense(input_dim=dim, units=V, kernel_initializer='uniform', activation= activation_extra_layer))

    skipgram.add( Dense(input_dim=dim, units=V, kernel_initializer='uniform', activation='softmax') )
    
    # Extra layer ()
    #
    # Justification: 
    #
    #
#     if extra_layer :
#         skipgram.add(Dense(input_dim=dim, units=V, kernel_initializer='uniform', activation= activation_extra_layer))
    
    # define loss function for Skipgram
    skipgram.compile(loss='categorical_crossentropy', optimizer='adadelta')

    return skipgram

In [17]:
################################################################################
########################## 2. Using Skipgram model #############################
################################################################################

dims = [50, 150, 300]
iterations = 10
listResX = []
listResY = []

for dimension in dims:
    print("Running Skipgram (dim = %s)..."% dimension)
    skipgram = getSkipgramModel(dimension, False)
    #print(skipgram.summary())
    X, Y = generate_data_skipgram(corpus, window_size, V)

    start_time = time.time()
    for ite in range(iterations):
        loss = 0.
        for x, y in zip(X,Y):
            loss += skipgram.train_on_batch(x, y)=
        print(ite, loss)
    end_time = time.time()
    
    print("--- %s seconds ---" % (end_time - start_time))
    listResX.append(X)
    listResY.append(Y)


Running Skipgram (dim = 50)...
0 41271.35240674019
1 39126.84181976318
2 39288.22808551788
3 39363.864663124084
4 39432.38656759262
5 39511.96853172779
6 39608.38287115097
7 39723.89285326004
8 39852.42831707001
9 39987.01795423031
--- 178.2564082145691 seconds ---
Running Skipgram (dim = 150)...
0 41222.66726255417
1 38923.405826091766
2 38988.34154701233
3 39005.65850484371
4 39038.850784897804
5 39093.706859230995
6 39156.13746762276
7 39216.62943792343
8 39267.45716762543
9 39305.05479335785
--- 347.22299909591675 seconds ---
Running Skipgram (dim = 300)...
0 41159.68823719025
1 38730.84329485893
2 38713.535183906555
3 38676.76882696152
4 38648.035522818565
5 38631.88415277004
6 38611.77810204029
7 38578.923778772354
8 38533.39666950703
9 38484.56395339966
--- 741.4052288532257 seconds ---


# CBOW additional hidden layer

In [18]:
################################################################################
################### 3. Using Skipgram model (extra layer) ######################
################################################################################

#dims = [50, 150, 300]
dims = [50]
iterations = 2
listResX = []
listResY = []
activation_rules = ["softmax", "elu", "selu", "softplus", "softsign", "relu", "tanh", "sigmoid", "hard_sigmoid", "linear"]
for dimension in dims:
    for activation_rule in activation_rules:
        print("Running Skipgram (dim = %s) with activation rule '%s'..."% (dimension, activation_rule))
        try:
            # set extra_layer = true
            cbow = getCbowModel(dimension, True, activation_rule)
            X, Y = generate_data_cbow(corpus, window_size, V)

            start_time = time.time()
            for ite in range(iterations):
                loss = 0.
                for x, y in zip(X,Y):
                    loss += cbow.train_on_batch(x, y)
                print(ite, loss)
            end_time = time.time()

            print("--- %s seconds ---" % (end_time - start_time))
            listResX.append(X)
            listResY.append(Y)

        except:
            print("Error with", activation_rule)
        

# Skipgram additional hidden layer

In [21]:
################################################################################
################### 3. Using Skipgram model (extra layer) ######################
################################################################################

#dims = [50, 150, 300]
dims = [50]
iterations = 2
listResX = []
listResY = []
activation_rules = ["softmax", "elu", "selu", "softplus", "softsign", "relu", "tanh", "sigmoid", "hard_sigmoid", "linear"]
for dimension in dims:
    for activation_rule in activation_rules:
        print("Running Skipgram (dim = %s) with activation rule '%s'..."% (dimension, activation_rule))
        try:
            # set extra_layer = true
            skipgram = getSkipgramModel(dimension, True, activation_rule)
            X, Y = generate_data_skipgram(corpus, window_size, V)

            start_time = time.time()
            for ite in range(iterations):
                loss = 0.
                for x, y in zip(X,Y):
                    loss += skipgram.train_on_batch(x, y)
                print(ite, loss)
            end_time = time.time()

            print("--- %s seconds ---" % (end_time - start_time))
            listResX.append(X)
            listResY.append(Y)

        except:
            print("Error with", activation_rule)
        

Running Skipgram (dim = 50) with activation rule 'softmax'...
0 41252.48093819618
1 39351.12027096748
--- 341.79421496391296 seconds ---
Running Skipgram (dim = 50) with activation rule 'elu'...
0 63386.02252089977
1 60039.88047675509
--- 354.2946639060974 seconds ---
Running Skipgram (dim = 50) with activation rule 'selu'...
0 71818.42933480814
1 69307.81007504463
--- 359.6227400302887 seconds ---
Running Skipgram (dim = 50) with activation rule 'softplus'...
0 41993.77337050438
1 39318.99882006645
--- 371.78749895095825 seconds ---
Running Skipgram (dim = 50) with activation rule 'softsign'...
0 64345.522713020444
1 64745.86871236283
--- 376.57004475593567 seconds ---
Running Skipgram (dim = 50) with activation rule 'relu'...
0 100396.4219110012
1 100394.24349093437
--- 374.63986110687256 seconds ---
Running Skipgram (dim = 50) with activation rule 'tanh'...
0 65003.985080762766
1 64692.410755110206
--- 1735.2997269630432 seconds ---
Running Skipgram (dim = 50) with activation rule '

In [22]:
################################################################################
####### 4a. Implement your own function to perform the analogy task with #######
################################################################################

embeddingWcoMatrix = "wcoMatrix"
# objA is to objB as objC is a objD
# we assume objs are strings
def analogyFunction(objA, objB, objC, objD, embeddingType = embeddingWcoMatrix, matrix = wcoMatrix):
    indexA = full_index[objA]
    indexB = full_index[objB]
    indexC = full_index[objC]
    indexD = full_index[objD]
    
    if(embeddingType == embeddingWcoMatrix):
        vecA = np.asarray(matrix[indexA,])
        vecB = np.asarray(matrix[indexB,])
        vecC = np.asarray(matrix[indexC,])
        vecD = np.asarray(matrix[indexD,])
    
    operationResult = vecA - vecB + vecD
    return ( np.array_equal(vecC,operationResult) ) # Not sure if this works

def makeAnalogyOnFile(embeddingType = embeddingWcoMatrix, matrix = wcoMatrix):
    # Read file
    content = []
    with open("analogy_alice.txt") as f:
        content = f.readlines()

    # Remove whitespace characters like '\n' at the end of each line
    content = [x.strip() for x in content] 
    allResults = []
    for analogy in content:
        words = analogy.split(" ")
        if len(words) == 4:
            try:
                result = analogyFunction(words[0], words[1], words[2], words[3])
                allResults.append(result)
                print("Testing '%s': %s" % (analogy, result))
            except KeyError:
                print("Missing words in the index ")
    return allResults

Comparison performance:

In [23]:
################################################################################
########### 4b. Compare the performance on the analogy task between ############
########### the word embeddings you have trained in 2.1, 2.2 & 2.3  ############
################################################################################


# 4. Analyze the four different word embeddings
#    - Compare the performance on the analogy task between the word embeddings that you have trained in 2.1, 2.2 and 2.3.  
#    - Visualize your results and interpret your results
# 5. Use the word co-occurence matrix from Question 1. Compare the performance on the analogy task with the performance of your trained word embeddings.  


analogyResultsWcoMatrix = makeAnalogyOnFile()
print("Score for w-w co-ocurrence matrix embedding: %s" % sum(analogyResultsWcoMatrix))

Missing words in the index 
Testing 'sudden suddenly usual usually': False
Testing 'bad worse good better': False
Testing 'go going look looking': False
Testing 'he she his her': False
Testing 'brother sister his her': False
Testing 'listen listening look looking': False
Testing 'saying said thinking thought': False
Testing 'bird birds cat cats': False
Testing 'good better old older': False
Testing 'good better quick quicker': False
Testing 'large largest good best': False
Missing words in the index 
Testing 'falling fell knowing knew': False
Testing 'walk walking think thinking': False
Testing 'child children cat cats': False
Testing 'dog dogs eye eyes': False
Testing 'hand hands rat rats': False
Testing 'eat eats find finds': False
Testing 'find finds say says': False
Testing 'old older good better': False
Testing 'large larger quick quicker': False
Testing 'go going listen listening': False
Testing 'run running walk walking': False
Testing 'run running think thinking': False
Missing

In [24]:
# Visualize results

# pylint: disable=missing-docstring
# Function to draw visualization of distance between embeddings.
def plot_with_labels(low_dim_embs, labels, filename):
  assert low_dim_embs.shape[0] >= len(labels), 'More labels than embeddings'
  plt.figure(figsize=(18, 18))  # in inches
  for i, label in enumerate(labels):
    x, y = low_dim_embs[i, :]
    plt.scatter(x, y)
    plt.annotate(label,
                 xy=(x, y),
                 xytext=(5, 2),
                 textcoords='offset points',
                 ha='right',
                 va='bottom')

  plt.savefig(filename)

try:
  # pylint: disable=g-import-not-at-top
  from sklearn.manifold import TSNE
  import matplotlib.pyplot as plt

  tsne = TSNE(perplexity=30, n_components=2, init='pca', n_iter=5000, method='exact')
  plot_only = 500
  low_dim_embs = tsne.fit_transform(final_embeddings[:plot_only, :])
  labels = [reverse_dictionary[i] for i in xrange(plot_only)]
  plot_with_labels(low_dim_embs, labels, os.path.join(gettempdir(), 'tsne.png'))

except ImportError as ex:
  print('Please install sklearn, matplotlib, and scipy to show embeddings.')
  print(ex)

NameError: name 'final_embeddings' is not defined

In [ ]:
################################################################################
############# 5. Use the word co-occurence matrix from Question 1. #############
################################################################################

# Compare the performance on the analogy task with the performance of your trained word embeddings.  
analogyResultsWcoMatrix = makeAnalogyOnFile()

print("\nScore for w-w co-ocurrence matrix embedding: %s" % sum(analogyResultsWcoMatrix))

Interpretation results of the visualization

Compare the results of the trained word embeddings with the word-word co-occurrence matrix

**6. Discuss**

Jorge > We can pick ideas from the paper and the book. Should be easy

#### What are the main advantages of CBOW and Skipgram?


#### What is the advantage of negative sampling?


#### What are the main drawbacks of CBOW and Skipgram?




7 Load pre-trained embeddings on large corpuses (see the pdf file). You only have to consider the word embeddings with an embedding size of 300

    - Compare performance on the analogy task with your own trained embeddings from "Alice in Wonderland". You can limit yourself to the vocabulary of Alice in Wonderland. Visualize the pre-trained word embeddings and compare these with the results of your own trained word embeddings. 


In [ ]:
#load pretrained word embeddings of word2vec
path_word2vec = "./GoogleNews-vectors-negative300.bin"
word2vec = KeyedVectors.load_word2vec_format(path_word2vec, binary=True)


In [ ]:
#load pretraind word embeddings of Glove

#path = "./glove.6B/glove.6B.300d_converted.txt"
path = "./glove.6B/glove.6B.300d.txt"

#convert GloVe into word2vec format
gensim.scripts.glove2word2vec.get_glove_info(path)
gensim.scripts.glove2word2vec.glove2word2vec(path, "glove_converted.txt")

glove = KeyedVectors.load_word2vec_format(path, binary=False)


In [ ]:
#Visualize the pre-trained word embeddings


Comparison performance with your own trained word embeddings